src/deterministic_vi/NewtonMaximize.jl

```diff
@@ -28,79 +32,101 @@ immutable Config{N,T}
+                         max_radius = max_radius)
 end
 
 #############
 # Objective #
 #############
 
-immutable Objective{F,N,T} <: Function
-    f::F
-    ea::ElboArgs{T}
-    f_evals::Ref{Int}
-    previous_x::Vector{T}
-    sf_free::SensitiveFloat{T}
-    cfg::Config{N,T}
+
+immutable Objective{N} <: Function
+    ea::ElboArgs
+    elbo_vars::ElboIntermediateVariables{Float64}
+    sf_free::SensitiveFloat{Float64}
+    cfg::Config{N,Float64}
+end
+
+immutable Gradient{N} <: Function
+    ea::ElboArgs
+    elbo_vars::ElboIntermediateVariables{Float64}
+    sf_free::SensitiveFloat{Float64}
+    cfg::Config{N,Float64}
+end
+
+immutable Hessian{N} <: Function
+    ea::ElboArgs
+    elbo_vars::ElboIntermediateVariables{Dual{1, Float64}}
+    sf_free::SensitiveFloat{Dual{1, Float64}}
+    cfg::Config{N,Dual{1, Float64}}
 end
 
-immutable Gradient{F<:Objective} <: Function
-    f::F
+"""
+An unconstrained version of the ELBO
+"""
+function Objective(ea::ElboArgs, cfg::Config, x::Vector)
+    elbo_vars = ElboIntermediateVariables(Float64, ea.S, length(ea.active_sources); 
+                                          calculate_gradient=calculate_gradient, 
+                                          calculate_hessian=false)
+    sf_free = SensitiveFloat{Float64}(length(cfg.free_params[1]),
+                                      length(cfg.bound_params), false, false)
+    return Objective(ea, elbo_vars, sf_free, cfg)
 end
 
-immutable Hessian{F<:Objective} <: Function
-    f::F
+"""
+Computes both the gradient the objective function's value
+"""
+function Gradient(ea::ElboArgs, cfg::Config, x::Vector)
+    elbo_vars = ElboIntermediateVariables(Float64, ea.S, length(ea.active_sources); 
+                                          calculate_gradient=calculate_gradient, 
+                                          calculate_hessian=false)
+    sf_free = SensitiveFloat{Float64}(length(cfg.free_params[1]),
+                                      length(cfg.bound_params), true, false)
+    return Gradient(ea, elbo_vars, sf_free, cfg)
 end
 
-function Objective{F,T}(f::F, ea::ElboArgs{T}, cfg::Config, x::Vector)
-    sf_free = SensitiveFloat{T}(length(cfg.free_params[1]), length(cfg.bound_params), true, true)
-    previous_x = fill(NaN, length(x))
-    return Objective(f, ea, Ref{Int}(0), previous_x, sf_free, cfg)
+"""
+Computes hessian-vector products
+"""
+function Hessian(ea::ElboArgs, cfg::Config, x::Vector)
+    elbo_vars = ElboIntermediateVariables(Dual{1, Float64}, ea.S, length(ea.active_sources); 
+                                          calculate_gradient=true, 
+                                          calculate_hessian=false)
+    sf_free = SensitiveFloat{Dual{1, Float64}}(length(cfg.free_params[1]), 
+                                    length(cfg.bound_params), true, false)
+    return Hessian(ea, elbo_vars, sf_free, cfg)
 end
 
 function evaluate!(f::Objective, x::Vector)
-    if x != f.previous_x
```

Comment:
Yes, that's it, the CG trust region code (not yet added to Optim) pretty much avoids calling the objective function twice with the same argument.